### PR TITLE
new packages: shotcut, mlt and their deps

### PR DESCRIPTION
**Pull request by knyipab**
Draft PR. currently the app mostly works ~well but cannot output video results, perhaps there are problems with QProcess, see #20354 .~ [fixed]

Other things to note:
* [fixed] ~32-bit arch does not build (see error log, related to `qint64` type)~
* [fixed] ~GPU mode requires  PR #20352 , related to #20351~
* [optional] rtaudio (mlt optional deps) needs ALSA and jack API, we may need a functioning pipewire, related to #20316 
* [optional: someone may try] `opencv` cannot be included as `mlt`build deps (which is done by Archlinux), prompting some compilation error related to `QPrivate` when building `shotcut`

### CODE REVIEW

File: x11-packages/shotcut/build.sh
Comment (TomJo2000) [inline x11-packages/shotcut/build.sh:15]: 32bit builds work when setting `-Wno-c++11-narrowing`, not sure if the resulting package is functional though.
```suggestion
"

termux_step_pre_configure(){
	CXXFLAGS+=' -Wno-c++11-narrowing'
}
```

**Comment by knyipab** [comment]
This package will be ready and I will force-push it once #20352 and #20374 are merged. 
The sound effects should work with sdl2 package. jack-dummy is more for runtime purpose (won't produce sound) otherwise need to patch the code. rtaudio is a longer run TODO item when pipewire is in a good state. opencv is nice-to-have item if someone find a fix but that shouldn't matter much.

**Comment by TomJo2000** [comment]
> This package will be ready and I will force-push it once #20352 and #20374 are merged. The sound effects should work with sdl2 package. jack-dummy is more for runtime purpose (won't produce sound) otherwise need to patch the code. rtaudio is a longer run TODO item when pipewire is in a good state. opencv is nice-to-have item if someone find a fix but that shouldn't matter much.

Please make sure to add
```bash
termux_step_pre_configure(){
	CXXFLAGS+=' -Wno-c++11-narrowing'
}
```
to make sure the 32 bit builds succeed.

**Comment by knyipab** [comment]
Thanks and will do when movit is available

**Comment by TomJo2000** [comment]
In that case I'm gonna go ahead and merge movit.

**Comment by knyipab** [comment]
Tested on my aarch64 phone. Import playlist, track editor, preview, some basic filters (rotate, opacity), export work.  

Ready to be merged. See if you are fine with the way I patch the source code.

**Comment by truboxl** [comment]
Please change all the .patch files that hardcoded with `/data/data/com.termux/files/usr` to `@TERMUX_PREFIX@`

**Comment by knyipab** [comment]
fixed. thanks

**Comment by Biswa96** [comment]
x11-packages/mlt/0001-cmake-mlt-undefined-version.patch and x11-packages/mlt/0002-cmake-mlt++-undefined-version.patch are probably not required. The issue can be fixed with `LDFLAGS+=" -Wl,--undefined-version"` (see packages/zlib/build.sh).

**Comment by knyipab** [comment]
Thanks @Biswa96 . You are right and force-pushed.

**Comment by truboxl** [comment]
Please change TERMUX_PKG_MAINTAINER field

**Comment by knyipab** [comment]
oh right, fixed.

File: packages/jack-dummy/build.sh
Comment (truboxl) [inline packages/jack-dummy/build.sh:13]: You dont need to manually `git submodule update`. Its done in `termux_git_clone_src` when `TERMUX_PKG_SRCURL` is a git repo. https://github.com/termux/termux-packages/blob/932cb3ea25991ed9020d1e4ca98c37b9e9f2a705/scripts/build/get_source/termux_git_clone_src.sh#L17

Also please remove the repeating `&&`.

File: x11-packages/mlt/0004-fhs-path.patch
Comment (truboxl) [inline x11-packages/mlt/0004-fhs-path.patch:75]: Can keep `$HOME/.frei0r-1/lib`

File: x11-packages/mlt/build.sh
Comment (truboxl) [inline x11-packages/mlt/build.sh:10]:
```suggestion
TERMUX_PKG_BUILD_DEPENDS="ffmpeg, fftw, fontconfig, frei0r-plugins, gdk-pixbuf, glib, jack-dummy, movit, ladspa-sdk, libepoxy, libexif, libglvnd, libsamplerate, libvidstab, libvorbis, libx11, libxml2, qt6-qtbase, qt6-qtsvg, pango, python, rubberband, sdl2, sdl-image, sox, zlib"
```

File: x11-packages/shotcut/build.sh
Comment (truboxl) [inline x11-packages/shotcut/build.sh:8]:
```suggestion
TERMUX_PKG_DEPENDS="ffmpeg, fftw, frei0r-plugins, mlt, libx264, libvpx, lame, ladspa-sdk, movit, qt6-qtbase, qt6-qtcharts, qt6-qtdeclarative, qt6-qtimageformats, qt6-qtmultimedia, qt6-qttranslations, sdl2"
```

File: x11-packages/mlt/0004-fhs-path.patch
Comment (knyipab) [reply x11-packages/mlt/0004-fhs-path.patch:75]: A functioning `jack2` and metapackage `jack` are on the way in PR #20544. I will wait for that and then remove this non-functioning and deprecated `jack1`. Will address your other suggestions after that. Thank you. 

Meanwhile, the jack library name `libjack.so.0` patch on `shotcut` will no longer be required after that.

**Comment by knyipab** [comment]
@truboxl refactored: addressed your comments, used the `jack` merged in #20544, removed TODO, and removed the libjack.so.0 patch.

**Comment by knyipab** [comment]
Just realized that this PR should be ready but has not yet merged. Any chance merging to provide `shotcut` to Termux users? And then close #20021

**Comment by TomJo2000** [comment]
LGTM.
Gonna wait for the CI to run through before merging.

<h1><em></em></h1> 

closes #20021

File: x11-packages/mlt/build.sh
Comment (TomJo2000) [inline x11-packages/mlt/build.sh:5]: Actually I just noticed, the build scripts are out of date now.
Would you mind bumping `mlt` to 7.28.0

File: x11-packages/shotcut/build.sh
Comment (TomJo2000) [inline x11-packages/shotcut/build.sh:5]: and `shotcut` to 24.09.13

**Review by TomJo2000 — CHANGES_REQUESTED** [review]

**Review by TomJo2000 — APPROVED** [review]
I went in and updated the build scripts.

I had to regenerate `mlt/0004-fhs-path.patch`, but that was just some line drift causing the 7.24.0 version of the patch to no longer apply.

**Comment by TomJo2000** [comment]
Guess that's what I get for only testing against AArch64 locally.
We're having symbol issues with 32Bit again.
```st
ld.lld: error: undefined symbol: Mlt::Producer::set_creation_time(long long)
>>> referenced by mltcontroller.cpp
>>>               src/CMakeFiles/shotcut.dir/mltcontroller.cpp.o:(Mlt::Controller::setupNewProducer(Mlt::Producer*) const)
>>> referenced by mltcontroller.cpp
>>>               src/CMakeFiles/shotcut.dir/mltcontroller.cpp.o:(Mlt::Controller::lockCreationTime(Mlt::Producer*) const)
>>> referenced by filedatedialog.cpp
>>>               src/CMakeFiles/shotcut.dir/dialogs/filedatedialog.cpp.o:(FileDateDialog::accept())
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by knyipab** [comment]
Related to https://github.com/mltframework/shotcut/issues/1580 and https://github.com/mltframework/mlt/issues/1020. 
TLDR: Someone suggested below patch and `mlt` dev said that only apply it conditionally for 32-bit build: 
```
--- a/src/mlt++/mlt++.vers
+++ b/src/mlt++/mlt++.vers
@@ -541,6 +541,7 @@ MLTPP_6.14.0 {
       "Mlt::Playlist::reorder(int const*)";
       "Mlt::Transition::connect(Mlt::Service&, int, int)";
       "Mlt::Producer::set_creation_time(long)";
+      "Mlt::Producer::set_creation_time(long long)";
       "Mlt::Producer::get_creation_time()";
   };
 } MLTPP_6.10.0;
```

**Comment by TomJo2000** [comment]
Should be easy enough.
I'll do it in a bit.

**Comment by TomJo2000** [comment]
`util-linux` has an example of a bit-ness depedent configure argument.
https://github.com/termux/termux-packages/blob/31e2ed440717d5f67d4e27bbea1a6617c2ba6a89/packages/util-linux/build.sh#L61-L63

We can use the same approach and just apply a `*.diff` patch instead.

**Comment by knyipab** [comment]
There seems to be a bug in the build system handling qt6 cross tool. Can't build even reverting to your version. 

Not many clues, I guess related to `gst-plugins-gl-headers` being installed during the build process for unknown reason and breaks other packages (it conflicts with mesa). Cuz I  encountered weird problem recently in TUR's on device building (mentioned in [this comment](https://github.com/termux-user-repository/tur/issues/1209#issuecomment-2366810129))

**Comment by truboxl** [comment]
The cross tools need to be rebuilt after the Ubuntu base image upgraded to 24.04.

**Comment by knyipab** [comment]
Thanks a lot! Ready to be merged.

**Review by TomJo2000 — APPROVED** [review]
LGTM.
Thanks a lot for your contribution and working with us on it.

**Pull request merged**
